keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+3. defaulting
+4. conversion to internal
+5. if immutability is enabled for the resource:
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized.
+
+### Future outline: protobuf
+
+Protobuf encoding unifies unset, empty and null values for slices and maps. Those three cases 
+cannot be differentiated an lead to the same on-the-wire value. In native types
+we use a semantic equality to factor those differences out if JSON is used on the wire, but
+protobuf is used for storage. In CRs decoding and encoding is always faithful with the 
+traditional JSON format used on-the-wire and for storage.
+
+When adding protobuf encoding to CRDs in the future, we have to (without major, non-standard
+efforts) identify unset, empty and null for CRs as well. This leads to the idea to use a less
+strict equality in the immutabiltiy checks with that case in mind. But protobuf encoding in
+native types also has a normalization effect, namely posted JSON object are normalized through
+the encoding to protobuf when writing to etcd. 
+
+Hence, it looks sensible to split normalization from immutability equality, and keep a strict 
+deep-equal equality even for protobuf, and potentially native types (if we decide to implement 
+immutability for those).
+
+With the container example we get this, with a sketch of a `+normalize` marker which
+normalizes empty and null to undefined:
+
+```
+// The list of containers can not change AT ALL after creation, modulo
+// empty, null and undefined.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+// +normalize=undefined
+Containers []Container `json:containers,omitempty` `protobuf:"bytes,2,opt,name=containers"
+```
+
+For fields which carry no `omitempty`, we could allow more advanced normalization
+modes which replicate the Golang serialization behaviour. Tooling like openapi-gen
+and the CRD validation could verify that the normalization specification matches 
+that of Golang.
+
+### Mutating admission chain
+
+Mutating admission chain would have the exact same effects as user changes,
+meaning that they wouldn't be able to change an object after creation. This is
+very similar to what is done today since validation for updates is run AFTER all
+mutations.
+
+### Where does this happen
+
+This process is meant to happen right before the update validation and after
+mutating, but before validating webhooks, and only run on updates. This will allow us to
+keep the exact same behavior while removing the validation code that checks the
+immutability of fields.
+
+![Decoding steps which must apply defaults](20191001-crd-immutability-pipeline.png)
+
+### Risks and Mitigations
+
+- immutable metadata would break API machinery. We forbid the `x-kubernetes-immutable`
```

Comment:
We could copy metadata from the built-in one. At first I thought maybe it'd be OK for users to add fields to it, but actually it's not safe--it would make it possible for a collision in the future if we add to metadata a field of the same name.

Reply:
Today we forbid any restriction to anything else than name and generated name. This is just a clarification that we don't change this with immutability.

Reply:
clarified.